### PR TITLE
[cxx] Port complex math to small portable C and C++.

### PR DESCRIPTION
**Pull request by jaykrell**
In file included from threadpool-worker-default.c:28:
../../mono/utils/mono-complex.h:67:8: warning: inline variables are a C++17 extension [-Wc++17-extensions]
static inline
       ^
../../mono/utils/mono-complex.h:68:15: error: expected ';' after top level declarator
double_complex mono_double_complex_make(gdouble re, gdouble im)
              ^
              ;

### CODE REVIEW

**Comment by jaykrell** [comment]
1. Concerns about C++ runtime dependency here? That's why I held back on Android but welcome more facts/opinions. I'm kinda operating in FUD territory on this one, and that is mostly ok, avoid the dependency, but not as easy here.
2. Does the "old" code compile asis on Android as C++? I need to try building the SDKs.
3. Anyone want to rewrite it w/o complex/imaginary?

**Comment by jaykrell** [comment]
Let's not commit this PR until more understood. It could take feedback/facts/opinions, but the present lack of facts is a bit unconvincing.

**Comment by jaykrell** [comment]
Maybe the right thing here is just leave this code as C?
To avoid any libc++/libstdc++ dependency?
And of course ensuring it interfaces with the rest of the world via C?

**Comment by jaykrell** [comment]
LGTM.

**Comment by jaykrell** [comment]
LGTM. Anyone else?

File: mono/utils/mono-complex.h [hunk 1]
```diff
@@ -60,38 +70,4 @@ double_complex mono_double_complex_sub(double_complex left, double_complex right
...
-}
-
-static inline
-double_complex mono_double_complex_sub(double_complex left, double_complex right)
-{
-	return left - right;
-}
-
-#endif
+#include "../support/libm/complex.c"
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/utils/mono-complex.h]: And separate file due to license. On one line of code. :(

File: mono/metadata/Makefile.am [hunk 2]
```diff
@@ -64,10 +64,6 @@ endif
 
 glib_libs = $(top_builddir)/mono/eglib/libeglib.la
 
-if HOST_ANDROID
```
Comment (lambdageek) [inline mono/metadata/Makefile.am]: I don't understand why we're getting rid of something that we apparently needed for android - and the commit message doesn't really explain it.

Comment (jaykrell) [reply mono/metadata/Makefile.am]: Commit message says "portable", i.e. one thing for all platfroms.

There were three functions there: `creal, `cimag`, `cabs`.
`creal`, `cimag` are in mono-complex.h now, one line each, inlined, `cabs` is left in this file which is not deleted, but included for inlining. Another choice is no inlining and remove the if android and build the file for all platforms. But one line, should be inlined.

File: mono/utils/mono-complex.h [hunk 3]
```diff
@@ -11,24 +11,34 @@
...
+	double_complex a = { re, im };
+	return a;
+}
+
+static inline double creal (double_complex c)
+{
+	return c.real;
+}
+
+static inline double cimag (double_complex c)
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/utils/mono-complex.h]: Could also be macros.

File: mono/metadata/Makefile.am (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/Makefile.am]: Is not clear enough?

**Comment by lambdageek** [comment]
Okay so in summary:
* We will never use the C99 `complex double` type.
* We will always use the `struct mono_double_complex` defined in `utils/mono-complex.h`
* We will redefine `double_complex` (our own thing) and `cabs`, `creal`, `cimag` (C99 standard names) to use the mono versions whenever the `utils/mono-complex.h` header is included.
* Previously on Android we used to use an implementation of complex math that looks like it was based on some FreeBSD code.  Now Android is no different than any other platform and we'll just use the `mono_double_complex` struct.

Is that accurate?  If yes, I'm ready to approve.

File: mono/utils/mono-complex.h [hunk 4]
```diff
@@ -11,24 +11,34 @@
...
-#endif
-
 #define _USE_MATH_DEFINES // needed by MSVC to define math constants
 #include <math.h>
 
-#ifdef _MSC_VER
+#define double_complex	mono_double_complex
+#define creal 		mono_double_complex_creal
+#define cimag 		mono_double_complex_cimag
+#define cabs 		mono_double_complex_cabs
```
Comment (lambdageek) [inline mono/utils/mono-complex.h]: We talked off-GitHub and @jaykrell will try to remove these #defines and update `threadpool-worker-default.c` to use `mono_double_complex_creal` etc.  Otherwise, LGTM.

**Comment by jaykrell** [comment]
So much line damage now, hard to read, let me try again.

**Comment by jaykrell** [comment]
I read some complex math lessons and they lined up with the preexisting msvc code but didn't check it all.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64
